db/multi_cf_iterator_test.cc

```diff
@@ -0,0 +1,198 @@
+
+  constexpr char key_3[] = "key_3";
+  WideColumns key_3_columns_in_cf_1{
+      {"cf_1_col_name_1", "cf_1_col_val_1_key_3"}};
+  WideColumns key_3_columns_in_cf_3{
+      {"cf_3_col_name_1", "cf_3_col_val_1_key_3"}};
+
+  constexpr char key_4[] = "key_4";
+  WideColumns key_4_columns_in_cf_0{
+      {"cf_0_col_name_1", "cf_0_col_val_1_key_4"}};
+  WideColumns key_4_columns_in_cf_2{
+      {"cf_2_col_name_1", "cf_2_col_val_1_key_4"}};
+
+  AttributeGroups key_1_attribute_groups{
+      AttributeGroup(handles_[2], key_1_columns_in_cf_2),
+      AttributeGroup(handles_[3], key_1_columns_in_cf_3)};
+  AttributeGroups key_2_attribute_groups{
+      AttributeGroup(handles_[1], key_2_columns_in_cf_1),
+      AttributeGroup(handles_[2], key_2_columns_in_cf_2)};
+  AttributeGroups key_3_attribute_groups{
+      AttributeGroup(handles_[1], key_3_columns_in_cf_1),
+      AttributeGroup(handles_[3], key_3_columns_in_cf_3)};
+  AttributeGroups key_4_attribute_groups{
+      AttributeGroup(handles_[0], key_4_columns_in_cf_0),
+      AttributeGroup(handles_[2], key_4_columns_in_cf_2)};
+
+  ASSERT_OK(db_->PutEntity(WriteOptions(), key_1, key_1_attribute_groups));
+  ASSERT_OK(db_->PutEntity(WriteOptions(), key_2, key_2_attribute_groups));
+  ASSERT_OK(db_->PutEntity(WriteOptions(), key_3, key_3_attribute_groups));
+  ASSERT_OK(db_->PutEntity(WriteOptions(), key_4, key_4_attribute_groups));
+
+  auto verify =
+      [&](const std::vector<ColumnFamilyHandle*>& cfhs,
+          const std::vector<Slice>& expected_keys,
+          const std::vector<Slice>& expected_values,
+          const std::vector<WideColumns>& expected_wide_columns,
+          const std::vector<AttributeGroups>& expected_attribute_groups) {
+        int i = 0;
+        MultiCfIterator* iter = db_->NewMultiCfIterator(ReadOptions(), cfhs);
+        for (iter->SeekToFirst(); iter->Valid(); iter->Next()) {
+          ASSERT_EQ(expected_keys[i], iter->key());
+          ASSERT_EQ(expected_values[i], iter->value());
+          ASSERT_EQ(expected_wide_columns[i], iter->columns());
+          ASSERT_EQ(expected_attribute_groups[i], iter->attribute_groups());
+          ++i;
+        }
+        delete iter;
+        // TODO - Check when implementation is done
+        // ASSERT_EQ(expected_keys.size(), i);
+      };
+
+  // Test for iteration over CF default->1->2->3
+  std::vector<ColumnFamilyHandle*> cfhs_order_0_1_2_3 = {
+      handles_[0], handles_[1], handles_[2], handles_[3]};
+  std::vector<Slice> expected_keys = {key_1, key_2, key_3, key_4};
+  // Pick what DBIter would return for value() in the first CF that key exists
+  // Since value for kDefaultWideColumnName only exists for key_1, rest will
+  // return empty value
+  std::vector<Slice> expected_values = {"cf_2_col_val_0_key_1", "", "", ""};
+
+  // Merge columns from all CFs that key exists and value is stored as wide
+  // column
+  std::vector<WideColumns> expected_wide_columns = {
```

Comment:
What happens when a column name is duplicated between CFs here? 

I originally thought `columns()` would behave like `value()` and returns WideColumns from the first CF that contains the key. And user can call `attribute_groups()` if they want results from all CFs. Is there a reason for returning this `expected_wide_columns`?

Reply:
That's a great question! I've briefly discussed this with @ltamasi. Please let me know what you think.

One idea that I suggested is providing an option to the user to choose the behavior.
1. Only the first one if key exists in multiple CFs
2. Merge them all (But this also has the similar issue - "same column name in multiple wide columns to merge")

However, I also feel that having more options can be confusing and against the "easier to use" vision we have.

For the default behavior, I now think that picking the first one only would be better as you suggested, to be consistent with what we return for `value()`. I'm open to more ideas / suggestions, though.